src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -305,22 +325,29 @@ bool IdealLoopTree::is_associative(Node* n, Node* base) {
 // (inv2 - x) - inv1  =>  (-inv1 + inv2) - x
 // inv1 - (x + inv2)  =>  ( inv1 - inv2) - x
 //
-Node* IdealLoopTree::reassociate_add_sub(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop *phase) {
-  assert(n1->is_Add() || n1->is_Sub(), "Target node should be add or subtract");
+// Apply the same transormations to == and !=
+// inv1 == (x + inv2) => ( inv1 - inv2 ) == x
+// inv1 == (x - inv2) => ( inv1 + inv2 ) == x
+// inv1 == (inv2 - x) => (-inv1 + inv2 ) == x
+//
+Node* IdealLoopTree::reassociate_add_sub_cmp(Node* n1, int inv1_idx, int inv2_idx, PhaseIdealLoop *phase) {
+  assert(n1->is_Add() || n1->is_Sub() || n1->is_Cmp(), "Target node should be add, subtract, or compare");
   Node* n2   = n1->in(3 - inv1_idx);
   Node* inv1 = n1->in(inv1_idx);
   Node* inv2 = n2->in(inv2_idx);
   Node* x    = n2->in(3 - inv2_idx);
 
-  bool neg_x    = n2->is_Sub() && inv2_idx == 1;
-  bool neg_inv2 = n2->is_Sub() && inv2_idx == 2;
-  bool neg_inv1 = n1->is_Sub() && inv1_idx == 2;
-  if (n1->is_Sub() && inv1_idx == 1) {
+  bool neg_x    = n2->is_Sub() && !n2->is_Cmp() && inv2_idx == 1;
+  bool neg_inv2 = (n2->is_Sub() && !n1->is_Cmp() && inv2_idx == 2) ||
+                  (n1->is_Cmp() && n2->is_Add());
+  bool neg_inv1 = (n1->is_Sub() && !n1->is_Cmp() && inv1_idx == 2) ||
+                  (n1->is_Cmp() && inv2_idx == 1 && n2->is_Sub());
+  if (n1->is_Sub() && !n1->is_Cmp() && inv1_idx == 1) {
```

Comment:
Would you mind adding some comments for this logic?

Reply:
I added a little comment block. Not sure how useful it is. I agree the code is hard to follow. Before all the changes for `Cmp`, I was able to make sense of it by following the comments at the top of the function.

Reply:
Hmm, how about using these variables?
```
bool n1_is_add = n1->is_Add();
bool n1_is_sub = n1->is_Sub() && !n1->is_Cmp();
bool n1_is_cmp = n1->is_Cmp();
```
Then you can just comment this:
`// Determine whether x, inv1, or inv2 should be negative in the transformed expression`